File: backend/app/models/matching.py
"""Pydantic models for answer matching endpoints."""
from typing import List, Optional
from pydantic import BaseModel


class MatchSettings(BaseModel):
    """User accommodation settings that control matching behavior."""
    match_acceptable_alternatives: bool = True
    match_partial_substring: bool = True
    match_word_overlap: bool = True
    match_stop_word_filtering: bool = True
    match_synonyms: bool = True
    match_first_name_only: bool = True


class NameMatchRequest(BaseModel):
    """Request to evaluate a name-practice answer."""
    user_answer: str
    expected_name: str
    nickname: Optional[str] = None
    alternatives: List[str] = []
    settings: Optional[MatchSettings] = None


class NameMatchResponse(BaseModel):
    """Result of a name-practice match."""
    is_correct: bool
    matched_via: Optional[str] = None  # exact, nickname, first_name, partial, etc.


class QuestionMatchRequest(BaseModel):
    """Request to evaluate a question-practice answer."""
    user_answer: str
    expected_answer: str
    acceptable_answers: List[str] = []
    settings: Optional[MatchSettings] = None


class QuestionMatchResponse(BaseModel):
    """Result of a question-practice match."""
    is_correct: bool
    is_partial: bool = False
    score: float = 0.0


class InformationMatchRequest(BaseModel):
    """Request to evaluate an information-practice answer."""
    user_answer: str
    expected_value: str
    field_name: str


class InformationMatchResponse(BaseModel):
    """Result of an information-practice match."""
    is_correct: bool
    confidence: float = 0.0


class WordFindingMatchRequest(BaseModel):
    """Request to evaluate a word-finding answer."""
    user_answer: str
    expected_name: str
    alternatives: List[str] = []


class WordFindingMatchResponse(BaseModel):
    """Result of a word-finding match."""
    is_correct: bool


class ExtractAnswerRequest(BaseModel):
    """Request to extract an answer from a speech transcript."""
    transcript: str


class ExtractAnswerResponse(BaseModel):
    """Extracted answer from speech."""
    answer: str

File: frontend/lib/api/matching.ts
/**
 * Typed helpers for the answer matching API endpoints.
 *
 * These are public endpoints (no auth needed) — they're called during
 * practice sessions and don't require user identity.
 */
import { apiClient } from './client'

// ============== Request/Response types ==============

export interface MatchSettings {
  match_acceptable_alternatives?: boolean
  match_partial_substring?: boolean
  match_word_overlap?: boolean
  match_stop_word_filtering?: boolean
  match_synonyms?: boolean
  match_first_name_only?: boolean
}

export interface NameMatchResult {
  is_correct: boolean
  matched_via: string | null
}

export interface QuestionMatchResult {
  is_correct: boolean
  is_partial: boolean
  score: number
}

export interface InformationMatchResult {
  is_correct: boolean
  confidence: number
}

export interface WordFindingMatchResult {
  is_correct: boolean
}

export interface ExtractAnswerResult {
  answer: string
}

// ============== API functions ==============

export async function evaluateNameAnswer(
  userAnswer: string,
  expectedName: string,
  options?: {
    nickname?: string
    alternatives?: string[]
    settings?: MatchSettings
  }
): Promise<NameMatchResult> {
  return apiClient.postPublic<NameMatchResult>('/api/matching/evaluate/name', {
    user_answer: userAnswer,
    expected_name: expectedName,
    nickname: options?.nickname,
    alternatives: options?.alternatives || [],
    settings: options?.settings,
  })
}

export async function evaluateQuestionAnswer(
  userAnswer: string,
  expectedAnswer: string,
  acceptableAnswers: string[] = [],
  settings?: MatchSettings
): Promise<QuestionMatchResult> {
  return apiClient.postPublic<QuestionMatchResult>('/api/matching/evaluate/question', {
    user_answer: userAnswer,
    expected_answer: expectedAnswer,
    acceptable_answers: acceptableAnswers,
    settings,
  })
}

export async function evaluateInformationAnswer(
  userAnswer: string,
  expectedValue: string,
  fieldName: string
): Promise<InformationMatchResult> {
  return apiClient.postPublic<InformationMatchResult>('/api/matching/evaluate/information', {
    user_answer: userAnswer,
    expected_value: expectedValue,
    field_name: fieldName,
  })
}

export async function evaluateWordFindingAnswer(
  userAnswer: string,
  expectedName: string,
  alternatives: string[] = []
): Promise<WordFindingMatchResult> {
  return apiClient.postPublic<WordFindingMatchResult>('/api/matching/evaluate/word-finding', {
    user_answer: userAnswer,
    expected_name: expectedName,
    alternatives,
  })
}

export async function extractAnswer(
  transcript: string
): Promise<string> {
  const result = await apiClient.postPublic<ExtractAnswerResult>('/api/matching/extract-answer', {
    transcript,
  })
  return result.answer
}
